File: scripts/render_cover.py
#!/usr/bin/env python3
"""Render an XML-safe cover SVG from a placeholder template."""

from datetime import date
from html import escape
from pathlib import Path
import re
import subprocess
import sys

ATTRIBUTE_TO_PLACEHOLDER = {
    "confidential-label": "CONFIDENTIAL_LABEL",
    "document-number": "DOCUMENT_NUMBER",
    "product-name": "PRODUCT_NAME",
    "cover-footer-text": "COVER_FOOTER_TEXT",
    "date": "FIELD_1_VALUE",
    "revision": "FIELD_2_VALUE",
}

GIT_HASH_SENTINEL = "GITHASH"
BUILD_DATE_SENTINEL = "BUILDDATE"
ENGLISH_MONTH_NAMES = (
    "",
    "January",
    "February",
    "March",
    "April",
    "May",
    "June",
    "July",
    "August",
    "September",
    "October",
    "November",
    "December",
)


def read_adoc_attributes(path: Path) -> dict[str, str]:
    attributes: dict[str, str] = {}
    for line in path.read_text(encoding="utf-8").splitlines():
        match = re.match(r"^:([a-z0-9_-]+):(?:\s+(.*))?$", line)
        if match:
            attributes[match.group(1)] = match.group(2) or ""
    return attributes


def resolve_revision(value: str, document: Path) -> str:
    if value != GIT_HASH_SENTINEL:
        return value
    try:
        revision = subprocess.run(
            ["git", "-C", str(document.parent), "rev-parse", "--short=12", "HEAD"],
            check=True,
            capture_output=True,
            text=True,
        ).stdout.strip()
        status = subprocess.run(
            [
                "git",
                "-C",
                str(document.parent),
                "status",
                "--porcelain",
                "--untracked-files=no",
            ],
            check=True,
            capture_output=True,
            text=True,
        ).stdout
    except (FileNotFoundError, subprocess.CalledProcessError) as error:
        raise SystemExit(
            "revision is GITHASH, but the Git commit hash could not be determined"
        ) from error
    return f"{revision}-dirty" if status else revision


def resolve_date(value: str) -> str:
    if value != BUILD_DATE_SENTINEL:
        return value
    build_date = date.today()
    return (
        f"{ENGLISH_MONTH_NAMES[build_date.month]} "
        f"{build_date.day}, {build_date.year}"
    )


def main() -> None:
    if len(sys.argv) != 5 or sys.argv[3] != "--adoc":
        raise SystemExit("usage: render_cover.py TEMPLATE OUTPUT --adoc DOCUMENT")

    source = Path(sys.argv[1]).read_text(encoding="utf-8")
    document = Path(sys.argv[4])
    attributes = read_adoc_attributes(document)
    attributes["date"] = resolve_date(attributes.get("date", ""))
    attributes["revision"] = resolve_revision(
        attributes.get("revision", ""), document
    )
    replacements = {
        "FIELD_1_LABEL": "DATE",
        "FIELD_2_LABEL": "REVISION",
    }
    replacements.update({
        placeholder: attributes.get(attribute, "")
        for attribute, placeholder in ATTRIBUTE_TO_PLACEHOLDER.items()
    })
    for key, value in replacements.items():
        source = source.replace(f"@{key}@", escape(value))
    Path(sys.argv[2]).write_text(source, encoding="utf-8")


if __name__ == "__main__":
    main()

File: scripts/render_cover.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

require "cgi"
require "date"
require "open3"

ATTRIBUTE_TO_PLACEHOLDER = {
  "confidential-label" => "CONFIDENTIAL_LABEL",
  "document-number" => "DOCUMENT_NUMBER",
  "product-name" => "PRODUCT_NAME",
  "cover-footer-text" => "COVER_FOOTER_TEXT",
  "date" => "FIELD_1_VALUE",
  "revision" => "FIELD_2_VALUE"
}.freeze

def read_adoc_attributes(path)
  File.readlines(path, chomp: true, encoding: "UTF-8").each_with_object({}) do |line, attributes|
    match = line.match(/^:([a-z0-9_-]+):(?:\s+(.*))?$/)
    attributes[match[1]] = match[2] || "" if match
  end
end

def git_output(document, *arguments)
  output, status = Open3.capture2e("git", "-C", File.dirname(document), *arguments)
  abort "revision is GITHASH, but the Git commit hash could not be determined" unless status.success?

  output
rescue Errno::ENOENT
  abort "revision is GITHASH, but Git is not installed"
end

def resolve_revision(value, document)
  return value unless value == "GITHASH"

  revision = git_output(document, "rev-parse", "--short=12", "HEAD").strip
  status = git_output(document, "status", "--porcelain", "--untracked-files=no")
  status.empty? ? revision : "#{revision}-dirty"
end

def resolve_date(value)
  return value unless value == "BUILDDATE"

  today = Date.today
  "#{Date::MONTHNAMES[today.month]} #{today.day}, #{today.year}"
end

def main
  unless ARGV.length == 4 && ARGV[2] == "--adoc"
    abort "usage: render_cover.rb TEMPLATE OUTPUT --adoc DOCUMENT"
  end

  template, output, _, document = ARGV
  source = File.read(template, encoding: "UTF-8")
  attributes = read_adoc_attributes(document)
  attributes["date"] = resolve_date(attributes.fetch("date", ""))
  attributes["revision"] = resolve_revision(attributes.fetch("revision", ""), document)
  replacements = {"FIELD_1_LABEL" => "DATE", "FIELD_2_LABEL" => "REVISION"}
  ATTRIBUTE_TO_PLACEHOLDER.each do |attribute, placeholder|
    replacements[placeholder] = attributes.fetch(attribute, "")
  end
  replacements.each do |placeholder, value|
    source = source.gsub("@#{placeholder}@", CGI.escapeHTML(value))
  end
  File.write(output, source, encoding: "UTF-8")
end

main if $PROGRAM_NAME == __FILE__
